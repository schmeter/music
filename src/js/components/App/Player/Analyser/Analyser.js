import React from 'react';
import PropTypes from 'prop-types';

import { isTouch } from '../../../../util/screen';

import configApp from '../../../../../config/app.json';

class Analyser extends React.Component {
    canvas = React.createRef();

    componentDidMount() {
        this.analyse();
    }

    componentDidUpdate() {
        this.analyse();
    }

    isAnalyserAllowed() {
        const { audio } = this.props;

        return audio && !isTouch() && configApp.useAnalyser;
    }

    analyse() {
        const { audio, isPlaying } = this.props;
        const canvas = this.canvas.current;
        const AudioContext = window.AudioContext;

        if (
            canvas
            && audio
            && AudioContext
            && !this.audioContext
            && isPlaying
        ) {
            this.audioContext = new AudioContext();
            const analyser = this.audioContext.createAnalyser();
            const source = this.audioContext.createMediaElementSource(audio);

            source.connect(analyser);
            analyser.connect(this.audioContext.destination);
            this.drawAnalyser(canvas, analyser);
        }
    }

    drawAnalyser(canvas, analyser) {
        const { mode = 'frequency' } = this.props;
        const bufferLength = analyser.frequencyBinCount;
        const dataArray = new Uint8Array(bufferLength);

        canvas.width = bufferLength;
        canvas.height = 256;
        const ctx = canvas.getContext('2d');
        const barWidth = 1;
        const barDistance = 1.375;
        const paintCanvas = () => {
            if (mode === 'waveform') {
                analyser.getByteTimeDomainData(dataArray);
            } else {
                analyser.getByteFrequencyData(dataArray);
            }
            ctx.clearRect(0, 0, canvas.width, canvas.height);
            dataArray.forEach((item, index) => {
                const barHeight = item * 3 / 4;

                if (mode === 'waveform') {
                    ctx.fillStyle = `rgba(255, 170, 0, 1)`;
                    ctx.fillRect(index, canvas.height - barHeight, barWidth, 1);
                } else {
                    ctx.fillStyle = `rgba(255, ${255 - item}, 0, ${1 - (item / canvas.height)})`;
                    ctx.fillRect(index * barDistance, canvas.height, barWidth, 0 - barHeight);
                }
            });
            window.requestAnimationFrame(paintCanvas);
        };

        paintCanvas();
    }

    render() {
        return !this.isAnalyserAllowed() ? null : (
            <canvas
                className="canvas"
                ref={this.canvas}
            />
        );
    }
}

Analyser.propTypes = {
    isPlaying: PropTypes.bool.isRequired,
    audio: PropTypes.object,
    mode: PropTypes.oneOf(['frequency', 'waveform'])
};

export default Analyser;
