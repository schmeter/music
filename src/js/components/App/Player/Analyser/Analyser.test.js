import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import Analyser from './Analyser';

window.AudioContext = class AudioContext {
    createAnalyser = () => ({
        connect: jest.fn(),
        getByteFrequencyData: jest.fn(),
        getByteTimeDomainData: jest.fn(),
        frequencyBinCount: 8
    });
    createMediaElementSource = () => ({
        connect: jest.fn()
    });
};

window.HTMLCanvasElement.prototype.getContext = () => ({
    clearRect: jest.fn(),
    fillRect: jest.fn()
});

it('renders correctly', () => {
    let props = {
        audio: {},
        isPlaying: false
    };

    let component = mount(<Analyser {...props} />);
    expect(toJson(component)).toMatchSnapshot();

    props = {
        ...props,
        isPlaying: true
    };

    component.setProps(props);
    component.update();
    expect(toJson(component)).toMatchSnapshot();

    component.unmount();

    props = {
        audio: {},
        isPlaying: false
    };

    component = mount(<Analyser {...props} />);
    expect(toJson(component)).toMatchSnapshot();

    props = {
        ...props,
        isPlaying: true,
        mode: 'waveform'
    };

    component.setProps(props);
    component.update();
    expect(toJson(component)).toMatchSnapshot();
});
