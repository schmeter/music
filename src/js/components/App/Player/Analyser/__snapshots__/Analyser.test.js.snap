// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Analyser
  audio={Object {}}
  isPlaying={false}
>
  <canvas
    className="canvas"
  />
</Analyser>
`;

exports[`renders correctly 2`] = `
<Analyser
  audio={Object {}}
  isPlaying={true}
>
  <canvas
    className="canvas"
  />
</Analyser>
`;

exports[`renders correctly 3`] = `
<Analyser
  audio={Object {}}
  isPlaying={false}
>
  <canvas
    className="canvas"
  />
</Analyser>
`;

exports[`renders correctly 4`] = `
<Analyser
  audio={Object {}}
  isPlaying={true}
  mode="waveform"
>
  <canvas
    className="canvas"
  />
</Analyser>
`;
